dev-docs/RFCs/v7.x/spring-based-transitions-rfc.md

```diff
@@ -69,13 +98,44 @@ new Layer({
 });
 ```
 
-[SAY SOME STUFF HERE ABOUT `onStart`, `onEnd`, AND `onInterrupt` CALLBACKS]
-
 Note the following change to the `enter` callback:
 
 The `enter` callback receives _one_ argument. A `toValue` (TypedArray) which represents the new value to transition to, for the current vertex. The `fromChunk` value that is supplied to `enter` for interpolation-based transitions is not supplied for spring-based transitions because no "from value" is maintained for the transition. 
 
 
 ## Implementation
 
-[FILL ME IN]
+A minimal spring-based transition system looks something like this:
+
+```js
+function createSpring (stiffness, damping, initialValue) {
+  let currentValue = initialValue
+  let previousValue = initialValue
+  return function tick (destinationValue) {
+    const velocity = currentValue - previousValue
+    const distanceToDestination = destinationValue - currentValue
+    const acceleration = distanceToDestination * stiffness - velocity * damping
+    const nextVal = currentValue + velocity + acceleration
+    previousValue = currentValue
+    currentValue = nextVal
+    return nextVal
+  }
+}
+```
+
+This requires the user to provide three values when defining the transition, `stiffness`, `damping`, and
+`initialValue` (i.e. `enter`), and a `destinationValue` on each tick. The library is then responsible for
+maintaining the current and previous values, from which it derives the velocity.
+
+For scalar uniforms, performing the transition on the CPU is as simple as the above example. Not much extra
+work is required to make this support vec2, vec3, and vec4 values. An `isAtDestination` function, which returns
+`true` when the `distance_to_destination` and `length(velocity)` are below some threshold, could be used to
+fire the `onEnd` callback and to stop the animation when the destination has been reached (though the
+calculation has such low overhead, I might be inclined to just run it on every frame).
+
+For attributes, the transition should be performed in a transform feedback pass. Three buffers are required to
+maintain the animating state: two read buffers for the `currentValue`s and `previousValue`s and one write buffer
+for the `nextValue`s. On each frame, the buffers are rotated. In other words: the `nextValue` buffers can become
+the `currentValue`, and the `currentValue` buffers can become the `previousValue`. This then frees up the
+`previousValue` buffers to be used as the write buffers for the `nextValue`s. The attribute buffers as they exist
+in the current implementation can become the `to` (or `destinationValue`) buffers.
```

Comment:
Is this just a clean extension to the existing system (additional parameters) or will the user have a choice between two animation systems?

Reply:
This would be a different transition system that extends the current transition API.

So the current (interpolation-based) transition system has us passing a `duration` param and (optionally) an `easing` param:
```js
  transitions: {
    getPositions: {
      duration: 300,
      easing: d3.easeCubicInOut
    }
```

And I'm proposing we let users define a (physics-based) spring transition by passing `stiffness` and `damping` params _instead_ of `duration` and `easing` params.

```js
  transitions: {
    getPositions: {
      spring: {
        stiffness: 0.001,
        damping: 0.05
      }
    }
```